### PR TITLE
change gear page size to 16 kb

### PR DESCRIPTION
**Pull request by grishasobol**
1) change gear page size to 16 kb
2) remove granularity page and psg case handling. Now granularity in storage is gear page.
3) refactoring lazy-pages so that they can work with dynamic page sizes - we can change page size in runtime upgrade.
4) global names, page sizes and storage prefix are set once pre block before process queue.
5) change some runtime interfaces to be more changeable and soft for changes from runtime upgrades.

### CODE REVIEW

**Review by ukint-vs — COMMENTED** [review]
Great

**Review by ukint-vs — APPROVED** [review]

File: common/lazy-pages/src/lib.rs [hunk 1]
```diff
@@ -44,29 +48,47 @@ fn mprotect_lazy_pages(mem: &mut impl Memory, protect: bool) {
...
+        weights.signal_write,
+        weights.signal_write_after_read,
+        weights.host_func_read,
+        weights.host_func_write,
+        weights.host_func_write_after_read,
+        weights.load_page_storage_data,
+    ]
+    .iter()
+    .map(|w| w.one())
+    .collect();
```
Comment (shamilsan) [inline common/lazy-pages/src/lib.rs:83]: No need to create an iterator for arrays.
```suggestion
    .map(|w| w.one())
    .to_vec();
```

File: common/lazy-pages/src/lib.rs [hunk 2]
```diff
@@ -105,27 +127,38 @@ pub fn update_lazy_pages_and_protect_again(
...
+        None
+    };
 
     let new_mem_size = mem.size();
-    if new_mem_size > old_mem_size {
-        gear_ri::set_wasm_mem_size(new_mem_size.into());
+    let changed_size = match new_mem_size > old_mem_size {
+        true => Some(new_mem_size.raw()),
+        false => None,
+    };
```
Comment (shamilsan) [inline common/lazy-pages/src/lib.rs:139]: One can simplify it.
```suggestion
    let changed_size = (new_mem_size > old_mem_size).then_some(new_mem_size.raw());
```

File: common/lazy-pages/src/lib.rs [hunk 3]
```diff
@@ -105,27 +127,38 @@ pub fn update_lazy_pages_and_protect_again(
...
 }
 
 /// Returns list of released pages numbers.
-pub fn get_released_pages() -> Vec<GearPage> {
-    gear_ri::get_released_pages()
+pub fn get_write_accessed_pages() -> Vec<GearPage> {
+    gear_ri::write_accessed_pages()
+        .iter()
+        .map(|p| {
+            GearPage::new(*p)
```
Comment (shamilsan) [inline common/lazy-pages/src/lib.rs:153]: We can consume the vector.
```suggestion
        .into_iter()
        .map(|p| {
            GearPage::new(p)
```

File: runtime-interface/src/lib.rs [hunk 4]
```diff
@@ -116,21 +92,28 @@ pub trait GearRI {
         (gas_left, res)
     }
 
-    fn get_lazy_pages_status() -> Option<Status> {
-        lazy_pages::get_status()
+    fn lazy_pages_status() -> (Status,) {
```
Comment (shamilsan) [inline runtime-interface/src/lib.rs:95]: Why not just `Status`?
```suggestion
    fn lazy_pages_status() -> Status {
```

File: lazy-pages/src/process.rs [hunk 5]
```diff
@@ -161,93 +123,87 @@ pub(crate) unsafe fn process_lazy_pages<H: AccessHandler>(
...
             }
         }
     };
 
-    let f = |pages: PagesIterInclusive<LazyPage>| {
-        let status = handler.charge_for_pages(&mut ctx, pages.clone())?;
-        if update_status(&mut ctx, status)? {
-            return Ok(());
-        }
+    // log::trace!("Process access from ")
```
Comment (shamilsan) [inline lazy-pages/src/process.rs:132]: Dead code?
```suggestion
```

File: common/lazy-pages/src/lib.rs (hunk 1 above)
Comment (grishasobol) [reply common/lazy-pages/src/lib.rs]: done

File: common/lazy-pages/src/lib.rs (hunk 2 above)
Comment (grishasobol) [reply common/lazy-pages/src/lib.rs]: done

File: lazy-pages/src/process.rs (hunk 5 above)
Comment (grishasobol) [reply lazy-pages/src/process.rs]: thank you) removed

File: lazy-pages/src/globals.rs [hunk 6]
```diff
@@ -18,15 +18,29 @@
...
-};
+use gear_backend_common::lazy_pages::{GlobalsAccessError, GlobalsAccessMod, GlobalsAccessor};
+use gear_core::memory::HostPointer;
 use sc_executor_common::sandbox::SandboxInstance;
 use sp_wasm_interface::Value;
 
-use crate::common::Error;
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum GlobalNo {
+    GasLimit = 0,
```
Comment (breathx) [inline lazy-pages/src/globals.rs:30]: enumerate all of them tho

File: lazy-pages/src/common.rs [hunk 7]
```diff
@@ -284,56 +265,34 @@ impl GasLeftCharger {
...
-        } else {
-            Ok(Status::Normal)
-        }
+    pub fn charge_for_page_data_load(&mut self, gas_left: &mut GasLeft) -> Status {
+        Self::sub_gas(gas_left, self.load_data_cost)
     }
 }
+
+pub(crate) enum WeightNo {
+    SignalRead = 0,
```
Comment (breathx) [inline lazy-pages/src/common.rs:290]: same

File: lazy-pages/src/globals.rs [hunk 8]
```diff
@@ -18,15 +18,29 @@
...
+    AllowanceLimit,
+    Amount,
```
Comment (breathx) [inline lazy-pages/src/globals.rs:32]: what's amount variant about?

File: pallets/gear/src/benchmarking/mod.rs [hunk 9]
```diff
@@ -989,7 +989,7 @@ benchmarks! {
     }
 
     gr_reply_push_input_per_kb {
-        let n in 0 .. T::Schedule::get().limits.payload_len / 1024;
+        let n in 0 .. MAX_PAYLOAD_LEN_KB / 1024;
```
Comment (breathx) [inline pallets/gear/src/benchmarking/mod.rs:992]:
```suggestion
        let n in 0 .. MAX_PAYLOAD_LEN_KB;
```

File: common/lazy-pages/src/lib.rs [hunk 10]
```diff
@@ -44,29 +48,46 @@ fn mprotect_lazy_pages(mem: &mut impl Memory, protect: bool) {
 }
 
 /// Try to enable and initialize lazy pages env
-pub fn try_to_enable_lazy_pages(pages_final_prefix: [u8; 32]) -> bool {
-    gear_ri::init_lazy_pages(pages_final_prefix)
+pub fn try_to_enable_lazy_pages(prefix: [u8; 32]) -> bool {
+    let ctx = LazyPagesRuntimeContext {
+        page_sizes: vec![WasmPage::size(), GearPage::size()],
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply common/lazy-pages/src/lib.rs:53]: if we will throw it as const size array then we will have problems with increasing it in future

File: lazy-pages/src/globals.rs (hunk 6 above)
Comment (grishasobol) [reply lazy-pages/src/globals.rs:30]: done

**Pull request merged**